### PR TITLE
WIP: Add create table syntax

### PR DESCRIPTION
**Pull request by nishikeshKardak**
Hello @klahnakoski, I am working on Issue #6. I have tried to convert SQL = "create table student (name varchar, age int )". but I can only convert into {'create table': [{'name': [['student']]}]}
For further SQL I can't understand how to convert it.

I have added WIP in the PR header because this PR is not completed. I'm currently working on it. Please review this and give me suggestions.
I have added the screenshot of the output.
Capture5

### CODE REVIEW

**Comment by klahnakoski** [comment]
Good! Good! 

You should be working from the finest detail, and build up from there. You must define what one column definition looks like. And to do that you must define the components:

    name VARCHAR2(80)

Think about what options you want to include (you need not do everything at first)

    'student_name' VARCHAR NOT NULL

So there are three parts here `name`, `type`, and `options`.  You will require a grammer for each

    column_options = Optional("NOT NULL")   # this can be improved

Then you can compose those three parts into a whole

    column_definition = column_name("name)+column_type("type)+column_options("options")

And then add that to your create table defintion:

    create_table = Group(
        ident.copy().setName("table_name")("name").setDebugActions(*debug)+
        Group(
            Literal("(").setDebugActions(*debug).suppress() +            
            delimitedList(column_definition) +
            Literal(")").setDebugActions(*debug).suppress()
        )("columns")
    )

I doubt my code works, but I think it will get you an idea of how to compose the grammar.

**Comment by nishikeshKardak** [comment]
Thanks, @klahnakoski, for very responsive feedback. I will make grammar as mentioned in the comment.

**Comment by nishikeshKardak** [comment]
Hello @klahnakoski, Can you please review this PL. I compose a grammar for creating table syntax and it works fine for me. I have attached a screenshot.
I test the sql query string  "create table student ( names varchar2, age int, rollNo int ) "
I did not add options for now. I will add soon.
#### Screenshot
Capture9

##### Thank you :)

**Comment by nishikeshKardak** [comment]
Hello @klahnakoski, I have resolved some grammar syntax problems. Please review this PL. 
I ran the test locally and I have the following results.

#### Screenshots
Capture7

**Comment by klahnakoski** [comment]
Please add some tests. You can add them to [simple.py](https://github.com/mozilla/moz-sql-parser/blob/dev/tests/test_simple.py).  Or, even better, make a new test file called `test_commands.py` that can hold these schema-changing statements.

**Comment by nishikeshKardak** [comment]
For now I'm working on column_options. After completing that,  I will make tests for these schema-changing statements.

**Comment by klahnakoski** [comment]
The code looks much like the code I gave you, so I have no feedback at this time.  I will require at tleast one test so I can run it through my debugger to see how it works.

**Comment by nishikeshKardak** [comment]
Hey @klahnakoski, The code you gave me gives me a clear idea about how to compose grammar using the pyparsing library.  That's why I worked on that. Now, I will be working on tests for the "create table syntax".

**Comment by nishikeshKardak** [comment]
While testing create table syntax, The SQL_PARSER is working fine. But it throws the following expectation while running tests.
```
C:\Users\Admin\Desktop\moz-sql-parser-dev\moz_sql_parser\__init__.py:12: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  from collections import Mapping
Traceback (most recent call last):
  File "C:\Users\Admin\Desktop\moz-sql-parser-dev\moz_sql_parser\__init__.py", line 104, in <module>
    print(parse("create table student (name varchar2, mnn nn )"))
  File "C:\Users\Admin\Desktop\moz-sql-parser-dev\moz_sql_parser\__init__.py", line 42, in parse
    parse_result = SQLParser.parseString(sql, parseAll=True)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1814, in parseString
    raise exc
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1804, in parseString
    loc, tokens = self._parse( instring, 0 )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 3886, in parseImpl
    raise maxException
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 3871, in parseImpl
    ret = e._parse( instring, loc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 4067, in parseImpl
    return self.expr._parse( instring, loc, doActions, callPreParse=False )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 3722, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 3722, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 4067, in parseImpl
    return self.expr._parse( instring, loc, doActions, callPreParse=False )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1548, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 3722, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1552, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 4067, in parseImpl
    return self.expr._parse( instring, loc, doActions, callPreParse=False )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1709, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 1530, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pyparsing.py", line 2625, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected {Forward: ... | Forward: ...} (at char 35), (line:1, col:36)
```

**Comment by nishikeshKardak** [comment]
The merge conflicts are due to some tests are failing. I have skipped them for now. I will add them after solving merge conflicts
```
ERROR: test_bad_select1 (test_simple.TestSimple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/mozilla/moz-sql-parser/tests/test_simple.py", line 111, in test_bad_select1
    assertRaises('Expected select', lambda: parse("se1ect A, B, C from dual"))
  File "tests/util.py", line 23, in assertRaises
    raise Exception("wrong error raised")
Exception: wrong error raised
======================================================================
ERROR: test_bad_select2 (test_simple.TestSimple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/mozilla/moz-sql-parser/tests/test_simple.py", line 114, in test_bad_select2
    assertRaises('Expected {{expression1 [{[as] column_name1}]}', lambda: parse("Select &&& FROM dual"))
  File "tests/util.py", line 23, in assertRaises
    raise Exception("wrong error raised")
Exception: wrong error raised
======================================================================
ERROR: test_incomplete1 (test_simple.TestSimple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/mozilla/moz-sql-parser/tests/test_simple.py", line 120, in test_incomplete1
    assertRaises('Expected {{expression1 [{[as] column_name1}]}', lambda: parse("SELECT"))
  File "tests/util.py", line 23, in assertRaises
    raise Exception("wrong error raised")
Exception: wrong error raised
----------------------------------------------------------------------
Ran 416 tests in 35.225s
FAILED (errors=3, skipped=19)
The command "python -m unittest discover tests" exited with 1.
```

**Comment by nishikeshKardak** [comment]
Fixed merged conflicts

**Comment by nishikeshKardak** [comment]
Hey @klahnakoski, Can you please review this PL. All the test for create table syntax are passing now. I resolve merge conflicts too.

**Comment by klahnakoski** [comment]
Alright! This is looking good!  

Please add a test for a column with some option (like NOT NULL)

File: moz_sql_parser/sql_parser.py [hunk 1]
```diff
@@ -77,6 +79,7 @@ def nothing(*args):
     "or",
     "order by",
     "select",
+    "table",
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: is this keyword used?

File: moz_sql_parser/sql_parser.py [hunk 2]
```diff
@@ -59,6 +59,8 @@ def nothing(*args):
     "between",
     "case",
     "collate nocase",
+    "columns",
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: is `columns` keyword used?

File: moz_sql_parser/sql_parser.py [hunk 3]
```diff
@@ -371,9 +375,71 @@ def to_string(instring, tokensStart, retTokens):
...
+    return {"columns" : tok.asList()}
+
+
+createStmt = Forward()
+
+column_name = ident.copy().setName("column_name").setDebugActions(*debug)
+
+column_size = Group(
+                Literal('(').suppress().setDebugActions(*debug) +
+                Word(nums) +
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: ` Word(nums)` is not correct; use `intNum` pattern

File: moz_sql_parser/sql_parser.py [hunk 4]
```diff
@@ -371,9 +375,71 @@ def to_string(instring, tokensStart, retTokens):
...
+                Literal(')').suppress().setDebugActions(*debug)
+              )
+
+column_type = ident.copy().setName("column_type").setDebugActions(*debug) + Optional(column_size("size"))
+
+column_options = Optional("NOT NULL") | Optional("UNIQUE")
+
+column_definition = Group(column_name("name") + column_type("type") + Optional(Literal(',').suppress()) + Optional(column_options("options")) + Optional(Literal(',').suppress())).addParseAction(to_create_json_call)
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: Use the `delimitedList` to handle comma-separated items

File: moz_sql_parser/sql_parser.py [hunk 5]
```diff
@@ -371,9 +375,71 @@ def to_string(instring, tokensStart, retTokens):
...
+
+
+createStmt << delimitedList(
+    CREATETABLE.suppress().setDebugActions(*debug) +
+    Group(delimitedList(
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: you do not need list `delimitedList`

File: moz_sql_parser/sql_parser.py (hunk 4 above)
Comment (klahnakoski) [reply moz_sql_parser/sql_parser.py]: in other words: The commas between column definitions should not be handled by the `column_definition`, do that with `delimitedList` in `createStmt`

File: moz_sql_parser/sql_parser.py [hunk 6]
```diff
@@ -371,9 +375,71 @@ def to_string(instring, tokensStart, retTokens):
...
+    ident.copy().setName("table_name")("name").setDebugActions(*debug)).addParseAction(to_table_name_call) +
+    Group(delimitedList(
+        Literal("(").setDebugActions(*debug).suppress() +
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: `delimitedList( Literal("(") ...` is attempting to parse a list of bracketed items.  This does not happen in create table; there is only one set of brackets.  Move the `Literal`s out of the `delimitedList()`

File: moz_sql_parser/sql_parser.py (hunk 2 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Yes Sir, We used this keyword for columns.
Example: `{'columns': [{'name': 'personid', 'type': 'int'}]`

File: moz_sql_parser/sql_parser.py (hunk 3 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: In intNum, we consider negative numbers also. So our grammar can accept negative numbers. That's why I choose only `nums`

File: moz_sql_parser/sql_parser.py (hunk 1 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: I will remove it

File: moz_sql_parser/sql_parser.py (hunk 3 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done!

File: moz_sql_parser/sql_parser.py (hunk 5 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done!

File: moz_sql_parser/sql_parser.py (hunk 6 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done!

File: moz_sql_parser/sql_parser.py (hunk 1 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done!

File: moz_sql_parser/sql_parser.py (hunk 4 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Completed!

**Comment by nishikeshKardak** [comment]
Het @klahnakoski, I done with all the changes and I have also added tests for column options.  All the tests are passing now. Can you please review this PL?

#### Screenshot
Capture9

Thanking you !

File: moz_sql_parser/sql_parser.py [hunk 7]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+
+createStmt << Group(
+    CREATETABLE.suppress().setDebugActions(*debug) +
+    Group(delimitedList(
+    ident.copy().setName("table_name")("name").setDebugActions(*debug)).addParseAction(to_table_name_call) +
+    Literal("(").setDebugActions(*debug).suppress() +
+    Group(
+        Group(OneOrMore(column_definition)).addParseAction(to_columns_call)
+        )("columns") +
+        Literal(")").setDebugActions(*debug).suppress()
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: Move this literal outside the `Group` too.  Consider the symmetry of `(` and `)`; they should be at the "same level" as each other; either both inside the `Group`  or both outside the group.

File: moz_sql_parser/sql_parser.py [hunk 8]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+    )
+
+column_options = Optional("NOT NULL") | Optional("UNIQUE")
+
+column_definition = delimitedList(
+        column_name("name") +
+        column_type("type") + Optional(Literal(',').suppress()) +
+        Optional(column_options("options")) +
+        Optional(Literal(',').suppress())
+    ).addParseAction(to_create_json_call)
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: This is looking nicer!, but it can be simplified more.  Remove the `Optional(Literal(',').suppress())` and use `delimitedList` instead of `OneOrMore` below.

File: moz_sql_parser/sql_parser.py [hunk 9]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+                Literal(')').suppress().setDebugActions(*debug)
+              )
+
+column_type = OneOrMore(
+    Group(
+        ident.copy().setName("column_type").setDebugActions(*debug) +
+        Optional(column_size("size")))
+    )
+
+column_options = Optional("NOT NULL") | Optional("UNIQUE")
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: Remove the `Optional()`, this is already assumed when you use or (`|`).  Use `ZeroOrMore` below in the main `createStmt`

File: moz_sql_parser/sql_parser.py [hunk 10]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+    return {"columns" : tok.asList()}
+
+
+createStmt = Forward()
+
+column_name = ident.copy().setName("column_name").setDebugActions(*debug)
+
+column_size = Group(
+                Literal('(').suppress().setDebugActions(*debug) +
+                intNum.copy().setName("size").setDebugActions(*debug) +
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: If I remember correctly, every pattern has two names.  The name used for parsing `setName("size")` and the name used for the output token `("size")`.  You should not need to `copy()`, rather 

    intNum("size").setDebugActions(*debug) +

Comment (klahnakoski) [reply moz_sql_parser/sql_parser.py]: yes, it can handle negative numbers.  Fix that if you want.

File: moz_sql_parser/sql_parser.py [hunk 11]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+        return None
+
+
+def to_create_table_call(instring, tokensStart, retTokens):
+    tok = retTokens
+
+    if tok:
+        return {"create table" : tok.asList()}
+
+def to_table_name_call(instring, tokensStart, retTokens):
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: You may not need this method; pyparsing should already be setting this for you.

File: moz_sql_parser/sql_parser.py [hunk 12]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+        Optional(Literal(',').suppress())
+    ).addParseAction(to_create_json_call)
+
+
+createStmt << Group(
+    CREATETABLE.suppress().setDebugActions(*debug) +
+    Group(delimitedList(
+    ident.copy().setName("table_name")("name").setDebugActions(*debug)).addParseAction(to_table_name_call) +
+    Literal("(").setDebugActions(*debug).suppress() +
+    Group(
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: With Literal(")") moved outside this `Group`, you will have two `Group`s, and you can remove one.

File: moz_sql_parser/sql_parser.py [hunk 13]
```diff
@@ -371,9 +374,104 @@ def to_string(instring, tokensStart, retTokens):
...
+        ident.copy().setName("column_type").setDebugActions(*debug) +
+        Optional(column_size("size")))
+    )
+
+column_options = Optional("NOT NULL") | Optional("UNIQUE")
+
+column_definition = delimitedList(
+        column_name("name") +
+        column_type("type") + Optional(Literal(',').suppress()) +
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: I do not think there are any commas (`,`) in a single column definition line.  Remove the `+ Optional(Literal(',').suppress()) `

File: moz_sql_parser/sql_parser.py (hunk 7 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done!

Comment (klahnakoski) [reply moz_sql_parser/sql_parser.py]: Excellent!  It is good to see you are still working on this!  I will look at it later today.

File: moz_sql_parser/sql_parser.py (hunk 8 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done !

File: moz_sql_parser/sql_parser.py (hunk 9 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Done !

Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Where can I use `ZeroOrMore` in the `createStmt`?? I have defined the `column_options` in `column_definition`.

File: moz_sql_parser/sql_parser.py (hunk 10 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Because of the user never puts negative integers in the `column_size`. So I replaced it with `intNum("size").setDebugActions(*debug)` and not fixing for positive integers.

File: moz_sql_parser/sql_parser.py (hunk 12 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Completed!

File: moz_sql_parser/sql_parser.py (hunk 13 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: Completed!

File: moz_sql_parser/sql_parser.py (hunk 11 above)
Comment (nishikeshKardak) [reply moz_sql_parser/sql_parser.py]: If I removed this method then I can't make the output string as expected. I have used this method for a column name. How should pyparsing already be setting this for us? Can you please tell me??

**Comment by klahnakoski** [comment]
Re-open this if you wish to get it passing tests
